### PR TITLE
semi-generic redi_push

### LINKED ISSUE #7: Discuss details for C/Fortran API
# Send one array
F: 

[i,d,f]redis_push(....)

C: 
`void Redis_push(redisConnnection * c, void * f, int dtype, int * dims, int ndims, char * KEY);`

PUBLISH subscription_key KEY

### PR DESCRIPTION
**Pull request by jhamman**
partially closes #7

cc @nbren12

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,18 +1,190 @@
+# CMAKE generated file: DO NOT EDIT!
```
Comment (nbren12) [inline Makefile]: Looks like the Makefile in the root got overwritten by cmake. The cmake builds should be run in `./build`. This is done automatically by running `make build` in the root directory.

File: src/redis_mod.f90 [hunk 2]
```diff
@@ -60,22 +71,83 @@ subroutine stream_data(f)
 
   end subroutine stream_data
 
-  subroutine redis_push(c, key, arr, dtype, dims)
```
Comment (nbren12) [inline src/redis_mod.f90]: I think we should probably leave this generic interface in place . Maybe we can change the name to `redis_push_explicit` or something, and then call it in the various `redis_push_f4` subroutines.

File: src/redis_mod.f90 [hunk 3]
```diff
@@ -2,6 +2,13 @@ module redis_mod
   use iso_c_binding
   implicit none
 
+! Test for a little-endian machine
```
Comment (nbren12) [inline src/redis_mod.f90]: Is the ` __BYTE_ORDER__ ` macro portable across different fortran compilers? Also, if we use macros, we should change the file extension from `.f90` to `.F90`.

File: test/test_redis_push.f90 [hunk 4]
```diff
@@ -3,24 +3,33 @@ program name
...
+      f8(i,j) = i + .5
+      i2(i,j) = i + 1
+      i4(i,j) = i + 1
     end do
   end do
 
   ndims = 2
 
-  call redis_push(redis, 'A', f, 'f4', shape(f))
+  call redis_push(redis, 'A-f4', f4)
```
Comment (nbren12) [inline test/test_redis_push.f90]: Cool. Now let's add some test for this in `test/test_send.py`.

File: CMakeLists.txt [hunk 5]
```diff
@@ -16,3 +16,5 @@ include_directories(${hiredis_INCLUDE_DIR})
 
 add_subdirectory (src)
 add_subdirectory (test)
+
+add_definitions(-cpp)
```
Comment (jhamman) [inline CMakeLists.txt]: is this the best way to add the C preprocessor to a cmake build?

**Review by nbren12 — COMMENTED** [review]
This looks awesome @jhamman. Checkout my comments.

File: CMakeLists.txt (hunk 5 above)
Comment (nbren12) [reply CMakeLists.txt]: I think if we change the extension from `.f90` to `.F90` the fortran compiler will know automatically to run the preprocessor.

**Comment by nbren12** [comment]
Overall, my preference would be to avoid using the preprocessor if possible and check the endianess using some of the solutions on this [stackoverflow thread](https://stackoverflow.com/questions/12791864/c-program-to-check-little-vs-big-endian).

**Comment by nbren12** [comment]
@jhamman Should we move this PR to a new branch like `feature/fortran-polymorphism`? That way I can push some changes.

**Comment by nbren12** [comment]
I just tried the to compile and I am getting amongst other errors.
```
 CALL redis_push(redis, 'A-f4', f4)
                                    1
Error: There is no specific subroutine for the generic 'redis_push' at (1)
```
This is happening because of the rank mismatch between f4, and the declared dimension of `arr` in `redis_push_f4(c, key, arr)`.

To make this work we would need to implement methods like `redis_push_{datatype}_{ndims}`. For example, if we want to implement this interface for real, float, and int arrays of up to 4 dimensions, it would require 12 different functions, which could ultimately be combined using an interface. I think this difficulty is why MPI asks user to specify manually specify a dtype and size when sending messages. 

IMO at this early point, I think it is probably better to keep the code base small, and ask the user to pass the dtype and shape. I do think we should check the endianess in the code, but I am pretty sure  `__BYTE_ORDER__` is a `gfortran` specific macro, so we should do this check at runtime.

**Comment by jhamman** [comment]
If you can give me another day, I think I'm pretty close on working out these kinks. I have to run to a meeting but I think I've gotten most of this figured out at this point.

**Comment by nbren12** [comment]
sure thing. take your time.

**Comment by jhamman** [comment]
@nbren12 - ready for a review. Not 100% satisfied with the API like this but it seems to be working.

File: src/redis_mod.F90 [hunk 6]
```diff
@@ -0,0 +1,455 @@
+MODULE redis_mod
+  USE iso_c_binding
```
Comment (jhamman) [inline src/redis_mod.F90]: Note: this file was renamed (suffix: `F90`) so its showing that all the lines were changed. I can revert this change if you want.

File: test/test_send.f90 [hunk 7]
```diff
@@ -1,19 +1,21 @@
...
+        f(i,j) = i
+     END DO
+  END DO
 
   ndims = 2
   dims = (/n,m/)
 
-  call array_to_redis(f)
-end program name
+  CALL array_to_redis(f)
```
Comment (jhamman) [inline test/test_send.f90]: I'm not really clear on what the purpose of this function is.

File: src/redis_mod.F90 (hunk 6 above)
Comment (nbren12) [reply src/redis_mod.F90]: we should go back to .f90 I think. This is causing a build error on travis, but probably not on your laptop bc mac is not case sensitive.

File: test/test_send.f90 (hunk 7 above)
Comment (nbren12) [reply test/test_send.f90]: I think this is the function we used to serve data to the bokeh server for the demonstration.

File: src/redis_wrapper.c [hunk 8]
```diff
@@ -3,196 +3,240 @@
...
+{
+    // dtype is in form <i4, <f8, <f4, etc
 
-  // get length of the array
-  char digit[2] = "";
-  digit[0] = dtype[1];
-  int bytesize = (size_t) atoi(digit);
-  size_t len = prod(bytesize, dims, ndims);
+    // get length of the array
+    char        digit[2] = "";
```
Comment (nbren12) [inline src/redis_wrapper.c]: This code, which determines the number of bytes from the dtype string, was broken by including the endianess indicator. You just have to change the index from 2 to 3. See this debugger output:
```
(lldb) n
Process 82666 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = step over
    frame #0: 0x000000010000464c test_redis_push`Redis_push(c=0x00000001005011e0, KEY="A-f4", f=0x00000001000050e0, dtype="<f4", dims=0x00007fff5fbff918, ndims=2) at redis_wrapper.c:218
   215 	    char        digit[2] = "";
   216 	    digit[0] = dtype[1];
   217 	    int         bytesize = (size_t)atoi(digit);
-> 218 	    size_t      len = prod(bytesize, dims, ndims);
   219
   220 	    // form dim string
   221 	    char        dimspec[100];
(lldb) p digit
(char [2]) $1 = "f"
```

File: test/test_connect.f90 [hunk 9]
```diff
@@ -1,11 +1,11 @@
-program name
-  use redis_mod
-  implicit none
+PROGRAM name
```
Comment (nbren12) [inline test/test_connect.f90:1]: Do you have a strong preference for upper-case statements? I usually just use lower-case everything in modern fortran code, as recommended [here](http://www.fortran90.org/src/best-practices.html). I just checked and WRF uses upper-case, but SAM and many other codes use lower-case.

File: geostreams/__init__.py [hunk 10]
```diff
@@ -4,10 +4,6 @@
 def get_test_exes():
```
Comment (nbren12) [inline geostreams/__init__.py]: I don't use this function anymore, so we can just delete it.

File: test/test_redis_push.f90 [hunk 11]
```diff
@@ -1,26 +1,42 @@
...
+        i2(i,j) = i + 1
+        i4(i,j) = i + 1
+     END DO
+  END DO
 
   ndims = 2
 
-  call redis_push(redis, 'A', f, 'f4', shape(f))
+  ! top level API
+  CALL redis_push(redis, 'A-f4', f4)
```
Comment (nbren12) [inline test/test_redis_push.f90:29]: These keys should each be tested in `test/test_send.py:test_redis_push`.

File: test/test_redis_push.f90 [hunk 12]
```diff
@@ -1,26 +1,42 @@
...
+  CALL redis_push(redis, 'A-f8', f8)
+  CALL redis_push(redis, 'A-i2', i2)
+  CALL redis_push(redis, 'A-i4', i4)
 
-  call free_connection(redis)
-end program name
+  ! Lower level API
+  CALL redis_push_explict(redis, 'A-f4', f4, '<f4', shape(f4))
```
Comment (nbren12) [inline test/test_redis_push.f90]: let's send change these keys to 'A-f4-explicit`, so we can also test them.

File: src/redis_mod.f90 [hunk 13]
```diff
@@ -1,81 +1,455 @@
...
+
+    ! array type
+    dtype = endian()//"i4"
+
+    ! Publish via explicit form
+    CALL redis_push_explict(c, key, arr, dtype, SHAPE(arr))
+  END SUBROUTINE redis_push_i4_3d
+
+
+  SUBROUTINE redis_push_explict(c, key, arr, dtype, dims)
```
Comment (nbren12) [inline src/redis_mod.f90]: typo in `explict`

File: src/redis_mod.f90 [hunk 14]
```diff
@@ -1,81 +1,455 @@
...
-    m = size(f, 2)
+    n = SIZE(f, 1)
+    m = SIZE(f, 2)
+
+    CALL publish_to_redis(f, n, m)
+
+  END SUBROUTINE stream_data
+
+
+  SUBROUTINE redis_push_f4_1d(c, key, arr)
```
Comment (nbren12) [inline src/redis_mod.f90]: I wish fortran had templates...sigh

**Review by nbren12 — COMMENTED** [review]
This is looking pretty good. Once we get the build working again by switching the file extension back to `.f90`, all we should have to do is add some testing code.

File: src/redis_wrapper.c [hunk 15]
```diff
@@ -3,196 +3,240 @@
...
+        redisCommand(c, "AUTH %s", pass);
+    }
 
-  return c;
+    redisCommand(c, "DEL A");
 
+    return c;
 }
 
-void publish_to_redis_(int* f, int* nptr , int* mptr){
```
Comment (nbren12) [inline src/redis_wrapper.c]: Is there a way to change the uncrusitfy settings so arguments and return types can stay on one line as long as that line is short enough?

Comment (jhamman) [reply src/redis_wrapper.c]: probably, I'll look into it.

File: test/test_send.f90 (hunk 7 above)
Comment (jhamman) [reply test/test_send.f90]: okay, I've removed this then.

File: test/test_redis_push.f90 (hunk 11 above)
Comment (jhamman) [reply test/test_redis_push.f90:29]: done.

**Comment by nbren12** [comment]
Moving to [feature/generic-interface](feature/generic-interface).

**Comment by nbren12** [comment]
I am merging this into a feature branch

**Pull request merged**